enhancements/cgu-precaching-additional-images.md

```diff
@@ -0,0 +1,308 @@
+`ClusterSelected -> Validated -> PrecacheSpecValid -> PrecachingSucceeded -> <TALM workflow continues>`.
+
+Detailed user and worklow steps are outlined below.
+
+1. User creates a CGU CR with the `preCaching` field set to `true` and prepares a ConfigMap (named
+`cluster-group-upgrade-overrides`) with the list of additional images to be pre-cached as shown in
+the examples below. Note that the ConfigMap should be created concurrently with the CGU or should
+exist prior to the creation of it. Furthermore, the ConfigMap should be created in the same namespace
+as the CGU on the hub cluster.
+   - Example CGU CR:
+    ```yaml
+    apiVersion: ran.openshift.io/v1alpha1
+    kind: ClusterGroupUpgrade
+    metadata:
+      name: cgu
+      namespace: default
+    spec:
+      clusters:
+      - sno1
+      - sno3
+      - sno7
+      enable: false
+      preCaching: true
+      backup: false
+      managedPolicies:
+      - policy0-common-config-policy
+      remediationStrategy:
+        maxConcurrency: 4
+        timeout: 240
+    ```
+   - Example ConfigMap:
+    ```yaml
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: cluster-group-upgrade-overrides
+      namespace: default
+    data:
+      additionalWorkloadImages: |
+        quay.io/foobar/application1@sha256:3d5800990dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47e2e1ef
+        quay.io/foobar/application2@sha256:3d5800123dee7cd4727d3fe238a97e2d2976d3808fc925ada29c559a47adfaef
+        ...
+        quay.io/foobar/applicationN@sha256:4fe1334adfafadsf987123adfffdaf1243340adfafdedga0991234afdadfsa09
+    ```
+
+2. The CR is validated (a function verifies if the clusters specified in the CR are part of a `ManagedCluster`
+object).
+
+3. If the validation succeeds, the pre-caching reconciliation commences. Note that all the sites are
+pre-cached concurrently.
+
+4. The pre-cache `spec` is then validated by checking if the managed policies exist. A valid `spec` will
+result in the following condition:
+    ```text
+    Type:    "PrecacheSpecValid",
+    Status:  True,
+    Reason:  "PrecacheSpecIsWellFormed",
+    Message: "Precaching spec is valid and consistent"
+    ```
+
+5. Pre-caching commences upon successful validation of the pre-cache `spec`. The following conditions are
+set while pre-caching is in-progress:
+    ```text
+    Type:    "PrecachingSuceeded",
+    Status:  False,
+    Reason:  "InProgress",
+    Message: "Precaching is required and not done"
+    ```
+
+6. After the pre-caching is successfully completed, the following condition is set:
+    ```text
+    Type:    "PrecachingSuceeded",
+    Status:  True,
+    Reason:  "PrecachingCompleted",
+    Message: "Precaching is completed for all clusters"
+    ```
+
+7. The remaining workflow for TALM continues as normal from here onwards.
+
+#### Variation [optional]
+
+N/A
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+
+N/A
```

Comment:
Any limits on number of additional images? 
Expectation of sufficient disk storage.
image spec must be fully qualified (full registry) with tag or sha or either?
You note this in the non-goals but it might be good to call out here that this is SNO only for now

Reply:
No limits on additional images or expectation of the disk storage. It is the responsibility of the user to ensure sufficient storage capacity proportional to the number of additional images to be pre-cached. 

I will add notes about the image specification as well as that this EP targets SNO-based architectures.

Reply:
Great suggestion regarding checking for GC and failing the pre-caching job accordingly. How about retrieving the `imageGCHighThresholdPercent` parameter from the cluster and checking the disk usage post pre-caching (i.e validate that the pre-caching job will not trigger the garbage collection)? However, GC will primarily impact the oldest and unused images first - so might not be worthwhile to set the pre-caching job as failed - rather we can capture this in the logs. I can certainly add this to the EP.

Reply:
@sudomakeinstall2 Made the same suggestion on post check when we did the pre check. @sakhoury Unfortunately, images we download is treated as oldest today as the last used timestamp is not set. We have another SPIKE for working with the CRIO team for either providing a way of pinning images or setting the timestamp. @vitus133 prototyped something that alters the timestamp by actually including the images in a dummy pod. But there are a bunch of issues associated with that approach.

Reply:
The above points have been captured in the `Implementation details/notes/constraints` section with the suggestions regarding how to check for disk space post pre-caching to avoid GC.